Make EXWM global key overrides customizable

timor opened the pull request:
If EXWM is mainly used in line-mode, it is useful to add the
bindings to the list of global prefix keys instead of setting them globally.
This way, they become accessible via the minor mode map, and can also be
deactivated again.

DamienCassou commented on README.org:
```diff
@@ -131,6 +131,22 @@ done by adding the following to your configuration file:
+**** EXWM Compatibility
+
+The customizable variable ~desktop-environment-update-exwm-global-keys~ can be
+used to to control how key bindings should be handled when EXWM is enabled.
```
"used to to control"

DamienCassou commented on desktop-environment.el:
```diff
@@ -411,12 +421,21 @@ the screen."
 (declare-function exwm-input-set-key "ext:exwm-input")
 
-(defun desktop-environment-exwm-set-global-keybindings ()
+(defun desktop-environment-exwm-set-global-keybindings (enable)
```
please add a reference to `ENABLE` in the function's docstring.

DamienCassou commented on desktop-environment.el:
```diff
@@ -176,6 +176,16 @@ portion of the screen."
+
+;;; Customization - EXWM keybindings
+
+(defcustom desktop-environment-update-exwm-global-keys :global
```
why `:global` and not `'global`? I never understood the real difference. If you have some pointer to the documentation, I will happily have a look.

timor replied:
For me, at least, it's a habit which mainly stems from Common Lisp.  The differences regarding namespace issues don't apply to elisp, but in general they are used wherever you would choose between a set of predefined options.

From https://www.gnu.org/software/emacs/manual/html_node/elisp/Symbol-Type.html:

> A symbol whose name starts with a colon (‘:’) is called a keyword symbol. These symbols automatically act as constants, and are normally used only by comparing an unknown symbol with a few specific alternatives.

DamienCassou commented on desktop-environment.el:
```diff
@@ -411,12 +421,21 @@ the screen."
-    (map-keymap (lambda (event definition)
-                  (exwm-input-set-key (vector event) definition))
-                desktop-environment-mode-map)))
+    (cond ((eq desktop-environment-update-exwm-global-keys :global)
```
you could use `pcase` or `cl-case` here if you want.

timor replied:
That would add dependencies, right?

DamienCassou replied:
Not sure what you mean exactly. Both are provided with Emacs so no external dependency is necessary. Both are autoloaded so no `require` form should be added. Additionally, `cl-case` doesn't even require `cl-lib` to be available at runtime.

timor replied:
Ah ok.  Will use `cl-case`, then.

DamienCassou commented on README.org:
```diff
@@ -131,6 +131,22 @@ done by adding the following to your configuration file:
+~desktop-environment-mode~, the bindings are deactivated again.
+
+Set to ~nil~ to disable any kind of special behavior in the presence of EXWM.
+
```
you may want to add that this variable only has an effect when EXWM is loaded.

timor replied:
Changed the wording in the first paragraph accordingly

DamienCassou commented on desktop-environment.el:
```diff
@@ -426,8 +445,8 @@ the screen."
   :lighter " DE"
-  (when desktop-environment-mode
-    (desktop-environment-exwm-set-global-keybindings)))
+  (when desktop-environment-update-exwm-global-keys
```
I think I would remove the `when` here to only keep the body and let `desktop-environment-exwm-set-global-keybindings` take care of `nil` itself.

DamienCassou commented on desktop-environment.el:
```diff
@@ -411,12 +421,28 @@ the screen."
+                                                      (cons event exwm-input-prefix-keys)
+                                                    (delq event exwm-input-prefix-keys))))
+                   desktop-environment-mode-map))
+      (nil nil)
```
According to my tests, I think this should be `((nil) nil)` instead. The documentation says:

> Each clause looks like (KEYLIST BODY...).  EXPR is evaluated and compared against each key in each KEYLIST; the corresponding BODY is evaluated.  If no clause succeeds, cl-case returns nil.  A **single non-nil atom** may be used in place of a KEYLIST of one atom.  A KEYLIST of t or otherwise is allowed only in the final clause, and matches if no other keys match.  Key values are compared by eql.

timor replied:
Of course. Will fix that.

DamienCassou commented on desktop-environment.el:
```diff
@@ -411,12 +421,28 @@ the screen."
+(defun desktop-environment-exwm-set-global-keybindings (enable)
+  "When using EXWM, add `desktop-environment-mode-map' to global keys if ENABLE
+  is non-nil.  When `desktop-environment-update-exwm-global-keys' is set to
+  :prefix, the bindings will be disabled when ENABLE is nil."
```
I think `:prefix' should be quoted.

DamienCassou commented:
I think your branch will trigger build failures with Travis, most probably from checkdoc. Please run checkdoc to make sure the new docstrings are ok.

timor commented:
Ok I think I addressed all issues.  Checkdoc gives one error, but that is unrelated to my changes.  Had to add a `(when (boundp ...))` guard around EXWM variable use to fix CI.
